### PR TITLE
LPS-159786 500 status error with FDS data

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-159786 500 status error with FDS data](https://issues.liferay.com/browse/LPS-159786)

### What is the goal of this PR?

In this PR, we are enabling FDS sample module to work in the CI environment. To achieve this, we are:
- reverting FDS data generation to execute on the first rendering, instead of module activation
- setting `companyId` to reflect currently used virtual instance, instead of the default virtual instance 
- enabling and fixing all previously disabled FDS tests.

In the current code, we are generating FDS sample data in the module activation phase, to the default virtual instance. Note that Objects data is tied to a virtual instance. This does not create sample data available in CI environment. CI creates a custom virtual instance, based on rather complex criteria. When the sample module is deployed, it is deployed on a page on the custom virtual instance. At the time of module activation, custom virtual instances don't exist, so there is no way to generate data in that phase. Having data generated on first render will always work, even if new virtual instance is created between each test.

### CODE REVIEW

**Comment by markocikos** [comment]
@john-co Looking at https://github.com/liferay-frontend/liferay-portal/pull/2763#issuecomment-1263656449, this is a valid failure. We changed the preloaded filter to be `Color`. In the test we:
1. Check `Pending` status option.
2. Uncheck `Approved` status option. This does nothing, as all three options are unchecked by default.
3. Test checks if we selected `Draft` and `Pending`. We didn't, so the test fails.

We need to update test.

But, this is good new, right? No 500 status error failure.

**Comment by john-co** [comment]
Just started reviewing :)

**Comment by john-co** [comment]
ci:test:frontend

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:stop

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
Rerunning tests with virtual instance and test fixes.

**Review by john-co — CHANGES_REQUESTED** [review]
tests still fail after enabling virtual instances

**Comment by markocikos** [comment]
@john-co Force-pushed, added a commit with `ModuleServiceLifecycle` change. I'm testing this because:
- I suspect the `ObjectDefinitionModel context class has not been injected` error has nothing to do with virtual instances. If it were, we would get some NPE in user or company service. This looks like an OSGi error, in wiring, referencing, service injecting or similar.
- Waiting for [all services to be available](https://help.liferay.com/hc/en-us/articles/360035828251-Waiting-on-Lifecycle-Events) seems like what the doctor ordered.
- This is set up in [Objects data generator](https://github.com/liferay/liferay-portal/blob/master/modules/apps/object/object-service/src/main/java/com/liferay/object/internal/ObjectDefinitionSampleGenerator.java#L176-L177).

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
NPEs in https://github.com/liferay-frontend/liferay-portal/pull/2763#issuecomment-1267282869 are different than before, it looks like we are past OSGi injection errors.

I added a bunch of `System.out.println`s in a temporary commit, to hopefully get some values in logs, to debug CI.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
CI does not work on draft PRs? :eyes:

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
Force-pushed, added more logging.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
Force-pushed, added even more logging

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
@dsanz @john-co  During the module activation, new VI does not exist yet. Debug logs prove this: 
1. There is only one data generation.
2. Later on, on page opening, new VI and user are taken from context.

```js
[beanshell] Generating FDS data:
[beanshell]  - userId = 20125
[beanshell]  - userId = test@liferay.com
[beanshell]  - companyId = 20097
[beanshell]  - companyVirtualHostname = localhost
[beanshell]  - objectDefinitionId = 43823

[beanshell] Getting entity model:
[beanshell]  - contextCompanyId = 44226
[beanshell]  - contextCompanyVirtualHostname = cloud-10-0-10-66.lax.liferay.com
[beanshell]  - contextUserId = 44273
[beanshell]  - contextUserEmailAddress = test@liferay.com
[beanshell]  - _objectDefinition is null
```
I don't see a way to resolve this. I'll move data generation on render.

**Comment by markocikos** [comment]
Force-pushed, with data generation on first render. There is a good chance that this does not work. It depends on how and when exactly does [`_objectDefinition`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/resource/v1_0/ObjectEntryResourceImpl.java#L503-L504) get initialized and injected. If this is done on request, code should work. If not, we'll get NPEs gain.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
Force-pushed. Looks like no related failures in https://github.com/liferay-frontend/liferay-portal/pull/2763#issuecomment-1271596563. With data generation on render, we can take `companyId` from request. No need to generate data in all VIs, no need for weird activation timing. I cleaned up code for this iteration, to make a final test. I believe the original 500 error issue from way back was simply because of the wrong `companyId`.

**Comment by markocikos** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/portlet/FDSSamplePortlet.java [hunk 1]
```diff
@@ -81,22 +82,22 @@ public void doDispatch(
...
-		super.doDispatch(renderRequest, renderResponse);
-	}
+		ThemeDisplay themeDisplay = (ThemeDisplay)renderRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
 
-	@Activate
-	protected void activate() {
 		try {
-			_generate(_portal.getDefaultCompanyId());
+			_generate(themeDisplay.getCompanyId());
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/portlet/FDSSamplePortlet.java:89]: back to the origins :) 
LGTM this looks much more safer for the stability of the testing even if that forces some delays in the first render

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
Before forwarding, I'd like to explain reasoning in the opening comment. Currently, it makes no sense with the code :sweat_smile:

**Review by john-co — APPROVED** [review]
Looks good to me! :sunglasses:

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**